File: breadth_first_search.py
from collections import deque, defaultdict

class BreadthFirstSearch:
    def __init__(self):
        self.graph = defaultdict(list)
    
    def add_edge(self, source, destination):
        self.graph[source].append(destination)
    
    def bfs(self, start_vertex):
        visited = set()
        queue = deque([start_vertex])
        visited.add(start_vertex)
        
        while queue:
            vertex = queue.popleft()
            print(vertex, end=" ")
            
            for adjacent in self.graph[vertex]:
                if adjacent not in visited:
                    visited.add(adjacent)
                    queue.append(adjacent)

if __name__ == "__main__":
    graph = BreadthFirstSearch()
    graph.add_edge(0, 1)
    graph.add_edge(0, 2)
    graph.add_edge(1, 2)
    graph.add_edge(2, 0)
    graph.add_edge(2, 3)
    graph.add_edge(3, 3)
    
    print("BFS traversal starting from vertex 2:")
    graph.bfs(2)

File: Java/graph/BreadthFirstSearch.java
import java.util.*;

public class BreadthFirstSearch {
    private int vertices;
    private LinkedList<Integer>[] adjacencyList;
    
    public BreadthFirstSearch(int vertices) {
        this.vertices = vertices;
        adjacencyList = new LinkedList[vertices];
        for (int i = 0; i < vertices; i++) {
            adjacencyList[i] = new LinkedList<>();
        }
    }
    
    public void addEdge(int source, int destination) {
        adjacencyList[source].add(destination);
    }
    
    public void bfs(int startVertex) {
        boolean[] visited = new boolean[vertices];
        Queue<Integer> queue = new LinkedList<>();
        
        visited[startVertex] = true;
        queue.add(startVertex);
        
        while (!queue.isEmpty()) {
            int vertex = queue.poll();
            System.out.print(vertex + " ");
            
            for (int adjacent : adjacencyList[vertex]) {
                if (!visited[adjacent]) {
                    visited[adjacent] = true;
                    queue.add(adjacent);
                }
            }
        }
    }
    
    public static void main(String[] args) {
        BreadthFirstSearch graph = new BreadthFirstSearch(4);
        graph.addEdge(0, 1);
        graph.addEdge(0, 2);
        graph.addEdge(1, 2);
        graph.addEdge(2, 0);
        graph.addEdge(2, 3);
        graph.addEdge(3, 3);
        
        System.out.println("BFS traversal starting from vertex 2:");
        graph.bfs(2);
    }
}